hedgehog/src/Hedgehog/Internal/State.hs

```diff
@@ -0,0 +1,474 @@
+  --   executed commands may have been removed from the sequence.
+  --
+    Require (state Symbolic -> input Symbolic -> Bool)
+
+  -- | Updates the model state, given the input and output of the command. Note
+  --   that this function is polymorphic in the type of values. This is because
+  --   it must work over 'Symbolic' values when we are generating actions, and
+  --   'Concrete' values when we are executing them.
+  --
+  | Update (forall v. Eq1 v => state v -> input v -> v output -> state v)
+
+  -- | A post-condition for a command that must be verified for the command to
+  --   be considered a success.
+  --
+  | Ensure (state Concrete -> input Concrete -> output -> Test m ())
+
+callbackRequire1 ::
+     state Symbolic
+  -> input Symbolic
+  -> Callback input output m state
+  -> Bool
+callbackRequire1 s i = \case
+  Require f ->
+    f s i
+  Update _ ->
+    True
+  Ensure _ ->
+    True
+
+callbackUpdate1 ::
+     Eq1 v
+  => state v
+  -> input v
+  -> v output
+  -> Callback input output m state
+  -> state v
+callbackUpdate1 s i o = \case
+  Require _ ->
+    s
+  Update f ->
+    f s i o
+  Ensure _ ->
+    s
+
+callbackEnsure1 ::
+     Monad m
+  => state Concrete
+  -> input Concrete
+  -> output
+  -> Callback input output m state
+  -> Test m ()
+callbackEnsure1 s i o = \case
+  Require _ ->
+    success
+  Update _ ->
+    success
+  Ensure f ->
+    f s i o
+
+callbackRequire ::
+     [Callback input output m state]
+  -> state Symbolic
+  -> input Symbolic
+  -> Bool
+callbackRequire callbacks s i =
+  all (callbackRequire1 s i) callbacks
+
+callbackUpdate ::
+     Eq1 v
+  => [Callback input output m state]
+  -> state v
+  -> input v
+  -> v output
+  -> state v
+callbackUpdate callbacks s0 i o =
+  foldl (\s -> callbackUpdate1 s i o) s0 callbacks
+
+callbackEnsure ::
+     Monad m
+  => [Callback input output m state]
+  -> state Concrete
+  -> input Concrete
+  -> output
+  -> Test m ()
+callbackEnsure callbacks s i o =
+  traverse_ (callbackEnsure1 s i o) callbacks
+
+------------------------------------------------------------------------
+
+-- | The specification for the expected behaviour of an 'Action'.
+--
+data Command n m (state :: (* -> *) -> *) =
+  forall input output.
+  (HTraversable input, Show (input Symbolic), Typeable output) =>
+  Command {
+    -- | A generator which provides random arguments for a command. If the
+    --   command cannot be executed in the current state, it should return
+    --   'Nothing'.
+    --
+      commandGen ::
+        state Symbolic -> Maybe (Gen n (input Symbolic))
+
+    -- | Executes a command using the arguments generated by 'commandGen'.
+    --
+    , commandExecute ::
+        input Concrete -> Test m output
+
+    -- | A set of callbacks which provide optional command configuration such
+    --   as pre-condtions, post-conditions and state updates.
+    --
+    , commandCallbacks ::
+        [Callback input output m state]
```

Comment:
As discussed, was just curious. As long as it's not just for convenience but because it's a better API, which is sounds like it is. Just ignore me. :)